### ISSUE TITLE
[MLv2] Implement functions for available fields for a given stage of the query

### ISSUE THREAD

**Issue by camsaul** [metabase/metabase#28691]
This is for powering various bits of the UI, e.g.

<img width="214" alt="image" src="https://user-images.githubusercontent.com/1455846/221763022-b6a284f0-dfe6-45c0-bb4e-012b8cc358ba.png">

Function might look something like

```clj
(visible-fields query stage-number) 
```

and

```clj
(visible-fields query) ; default to stage -1 (last stage) 
```

Not sure about the name.

This can probably just take the `:lib/stage-metadata` for a stage of a query (this is added by #28717) and return that (the JS version would need to do `clj->js` on the maps before returning them). We might need to add some extra info not already in the metadata -- TBD. Figuring out what we need to change if anything is part of this task

**Comment by camsaul** [comment]
Actually `:lib/stage-metadata` will only kind of work, because if we already have `:fields` then presumably the metadata will be updated to only include those Fields, and the UI needs to show us Fields that are not currently included as well. So maybe we'll have to look at the metadata from the previous stage or the source Table to see the complete list of everything that can be added.

**Comment by camsaul** [comment]
The response needs to be JS/UI friendly (probably an array of maps), and we need to be able to take one or more of the maps returned by that response and pass it in to `(fields query ...)` to set the `:fields` for the query

**Comment by camsaul** [comment]
We need to include implicitly joinable fields for several places as well. But not all of them! Maybe this should be an option.

<img width="252" alt="image" src="https://user-images.githubusercontent.com/1455846/228046865-635a50eb-2a07-453b-a411-978f6fce5ee0.png">

<img width="266" alt="image" src="https://user-images.githubusercontent.com/1455846/228046917-07267ea0-3ade-4765-97d9-22267b1dd7d6.png">

<img width="273" alt="image" src="https://user-images.githubusercontent.com/1455846/228046950-f0041a83-e5bd-4536-b209-ff103b3990ad.png">

<img width="263" alt="image" src="https://user-images.githubusercontent.com/1455846/228047006-e2a71573-e4f9-4169-a263-b1374f1bf305.png">

**Comment by camsaul** [comment]
Example `GET /api/table/:id/query_metadata` response:

```json
{
  "description": "Includes a catalog of all the products ever sold by the famed Sample Company.",
  "entity_type": "entity/ProductTable",
  "schema": "PUBLIC",
  "db": {
    "description": "Some example data for you to play around with as you embark on your Metabase journey.",
    "features": [
      "basic-aggregations",
      "temporal-extract",
      "actions/custom",
      "now",
      "standard-deviation-aggregations",
      "date-arithmetics",
      "actions",
      "expression-aggregations",
      "foreign-keys",
      "right-join",
      "left-join",
      "native-parameters",
      "nested-queries",
      "expressions",
      "regex",
      "case-sensitivity-string-filter-options",
      "binning",
      "datetime-diff",
      "inner-join",
      "advanced-math-expressions"
    ],
    "cache_field_values_schedule": "0 0 17 * * ? *",
    "timezone": "UTC",
    "auto_run_queries": true,
    "metadata_sync_schedule": "0 57 * * * ? *",
    "name": "Sample Database",
    "settings": null,
    "caveats": "You probably don't want to use this for your business-critical analyses, but hey, it's your world, we're just living in it.",
    "creator_id": null,
    "is_full_sync": true,
    "updated_at": "2023-03-23T18:04:15.38504Z",
    "cache_ttl": null,
    "details": null,
    "is_sample": true,
    "id": 2177,
    "is_on_demand": false,
    "options": null,
    "engine": "h2",
    "initial_sync_status": "complete",
    "dbms_version": {
      "flavor": "H2",
      "version": "2.1.214 (2022-06-13)",
      "semantic-version": [
        2,
        1
      ]
    },
    "refingerprint": null,
    "created_at": "2023-03-23T18:04:15.38504Z",
    "points_of_interest": "You can find all sorts of different joinable tables ranging from products to people to reviews here."
  },
  "show_in_getting_started": false,
  "name": "PRODUCTS",
  "fields": [
    {
      "description": "The numerical product number. Only used internally. All external communication should use the title or EAN.",
      "database_type": "BIGINT",
      "semantic_type": "type/PK",
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "ID",
      "fingerprint_version": 0,
      "has_field_values": "none",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2023-03-23T18:04:15.712051Z",
      "custom_position": 0,
      "effective_type": "type/BigInteger",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45479,
      "last_analyzed": null,
      "database_is_auto_increment": true,
      "position": 0,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "ID",
      "database_position": 0,
      "database_required": false,
      "fingerprint": null,
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/BigInteger",
      "points_of_interest": null
    },
    {
      "description": "The international article number. A 13 digit number uniquely identifying the product.",
      "database_type": "CHARACTER",
      "semantic_type": null,
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "EAN",
      "fingerprint_version": 5,
      "has_field_values": "list",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2023-03-23T18:04:17.216127Z",
      "custom_position": 0,
      "effective_type": "type/Text",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45483,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 1,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "Ean",
      "database_position": 1,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 200,
          "nil%": 0
        },
        "type": {
          "type/Text": {
            "percent-json": 0,
            "percent-url": 0,
            "percent-email": 0,
            "percent-state": 0,
            "average-length": 13
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/Text",
      "points_of_interest": null
    },
    {
      "description": "The name of the product as it should be displayed to customers.",
      "database_type": "CHARACTER VARYING",
      "semantic_type": "type/Title",
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "TITLE",
      "fingerprint_version": 5,
      "has_field_values": "list",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2023-03-23T18:04:17.210883Z",
      "custom_position": 0,
      "effective_type": "type/Text",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45478,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 2,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "Title",
      "database_position": 2,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 199,
          "nil%": 0
        },
        "type": {
          "type/Text": {
            "percent-json": 0,
            "percent-url": 0,
            "percent-email": 0,
            "percent-state": 0,
            "average-length": 21.495
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/Text",
      "points_of_interest": null
    },
    {
      "description": "The type of product, valid values include: Doohicky, Gadget, Gizmo and Widget",
      "database_type": "CHARACTER VARYING",
      "semantic_type": "type/Category",
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "CATEGORY",
      "fingerprint_version": 5,
      "has_field_values": "list",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2023-03-23T18:04:17.21455Z",
      "custom_position": 0,
      "effective_type": "type/Text",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45481,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 3,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "Category",
      "database_position": 3,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 4,
          "nil%": 0
        },
        "type": {
          "type/Text": {
            "percent-json": 0,
            "percent-url": 0,
            "percent-email": 0,
            "percent-state": 0,
            "average-length": 6.375
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/Text",
      "points_of_interest": null
    },
    {
      "description": "The source of the product.",
      "database_type": "CHARACTER VARYING",
      "semantic_type": "type/Company",
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "VENDOR",
      "fingerprint_version": 5,
      "has_field_values": "list",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [],
      "updated_at": "2023-03-23T18:04:17.212783Z",
      "custom_position": 0,
      "effective_type": "type/Text",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45480,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 4,
      "visibility_type": "normal",
      "default_dimension_option": null,
      "target": null,
      "preview_display": true,
      "display_name": "Vendor",
      "database_position": 4,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 200,
          "nil%": 0
        },
        "type": {
          "type/Text": {
            "percent-json": 0,
            "percent-url": 0,
            "percent-email": 0,
            "percent-state": 0,
            "average-length": 20.6
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/Text",
      "points_of_interest": null
    },
    {
      "description": "The list price of the product. Note that this is not always the price the product sold for due to discounts, promotions, etc.",
      "database_type": "DOUBLE PRECISION",
      "semantic_type": null,
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "PRICE",
      "fingerprint_version": 5,
      "has_field_values": "none",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [
        "29",
        "30",
        "31",
        "32",
        "33"
      ],
      "updated_at": "2023-03-23T18:04:17.048794Z",
      "custom_position": 0,
      "effective_type": "type/Float",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45477,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 5,
      "visibility_type": "normal",
      "default_dimension_option": "29",
      "target": null,
      "preview_display": true,
      "display_name": "Price",
      "database_position": 5,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 170,
          "nil%": 0
        },
        "type": {
          "type/Number": {
            "min": 15.691943673970439,
            "q1": 37.25154462926434,
            "q3": 75.45898071609447,
            "max": 98.81933684368194,
            "sd": 21.711481557852057,
            "avg": 55.74639966792074
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/Float",
      "points_of_interest": null
    },
    {
      "description": "The average rating users have given the product. This ranges from 1 - 5",
      "database_type": "DOUBLE PRECISION",
      "semantic_type": "type/Score",
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "RATING",
      "fingerprint_version": 5,
      "has_field_values": "none",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [
        "29",
        "30",
        "31",
        "32",
        "33"
      ],
      "updated_at": "2023-03-23T18:04:17.219709Z",
      "custom_position": 0,
      "effective_type": "type/Float",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45484,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 6,
      "visibility_type": "normal",
      "default_dimension_option": "29",
      "target": null,
      "preview_display": true,
      "display_name": "Rating",
      "database_position": 6,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 23,
          "nil%": 0
        },
        "type": {
          "type/Number": {
            "min": 0,
            "q1": 3.5120465053408525,
            "q3": 4.216124969497314,
            "max": 5,
            "sd": 1.3605488657451452,
            "avg": 3.4715
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/Float",
      "points_of_interest": null
    },
    {
      "description": "The date the product was added to our catalog.",
      "database_type": "TIMESTAMP",
      "semantic_type": "type/CreationTimestamp",
      "table_id": 9807,
      "coercion_strategy": null,
      "name": "CREATED_AT",
      "fingerprint_version": 5,
      "has_field_values": "none",
      "settings": null,
      "caveats": null,
      "fk_target_field_id": null,
      "dimensions": [],
      "dimension_options": [
        "11",
        "12",
        "13",
        "14",
        "15",
        "16",
        "17",
        "18",
        "19",
        "20",
        "21",
        "22",
        "23",
        "24",
        "25"
      ],
      "updated_at": "2023-03-23T18:04:17.217988Z",
      "custom_position": 0,
      "effective_type": "type/DateTime",
      "active": true,
      "nfc_path": null,
      "parent_id": null,
      "id": 45482,
      "last_analyzed": "2023-03-23T18:04:17.267596Z",
      "database_is_auto_increment": false,
      "position": 7,
      "visibility_type": "normal",
      "default_dimension_option": "13",
      "target": null,
      "preview_display": true,
      "display_name": "Created At",
      "database_position": 7,
      "database_required": false,
      "fingerprint": {
        "global": {
          "distinct-count": 200,
          "nil%": 0
        },
        "type": {
          "type/DateTime": {
            "earliest": "2016-04-26T19:29:55.147Z",
            "latest": "2019-04-15T13:34:19.931Z"
          }
        }
      },
      "created_at": "2023-03-23T18:04:15.712051Z",
      "base_type": "type/DateTime",
      "points_of_interest": null
    }
  ],
  "caveats": "The rating column is an integer from 1-5 where 1 is dreadful and 5 is the best thing ever.",
  "segments": [],
  "dimension_options": {
    "0": {
      "name": "Day",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day"
        }
      ],
      "type": "type/Date"
    },
    "1": {
      "name": "Week",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "week"
        }
      ],
      "type": "type/Date"
    },
    "2": {
      "name": "Month",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "month"
        }
      ],
      "type": "type/Date"
    },
    "3": {
      "name": "Quarter",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "quarter"
        }
      ],
      "type": "type/Date"
    },
    "4": {
      "name": "Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "year"
        }
      ],
      "type": "type/Date"
    },
    "5": {
      "name": "Day of Week",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day-of-week"
        }
      ],
      "type": "type/Date"
    },
    "6": {
      "name": "Day of Month",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day-of-month"
        }
      ],
      "type": "type/Date"
    },
    "7": {
      "name": "Day of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day-of-year"
        }
      ],
      "type": "type/Date"
    },
    "8": {
      "name": "Week of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "week-of-year"
        }
      ],
      "type": "type/Date"
    },
    "9": {
      "name": "Month of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "month-of-year"
        }
      ],
      "type": "type/Date"
    },
    "10": {
      "name": "Quarter of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "quarter-of-year"
        }
      ],
      "type": "type/Date"
    },
    "11": {
      "name": "Minute",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "minute"
        }
      ],
      "type": "type/DateTime"
    },
    "12": {
      "name": "Hour",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "hour"
        }
      ],
      "type": "type/DateTime"
    },
    "13": {
      "name": "Day",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day"
        }
      ],
      "type": "type/DateTime"
    },
    "14": {
      "name": "Week",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "week"
        }
      ],
      "type": "type/DateTime"
    },
    "15": {
      "name": "Month",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "month"
        }
      ],
      "type": "type/DateTime"
    },
    "16": {
      "name": "Quarter",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "quarter"
        }
      ],
      "type": "type/DateTime"
    },
    "17": {
      "name": "Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "year"
        }
      ],
      "type": "type/DateTime"
    },
    "18": {
      "name": "Minute of Hour",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "minute-of-hour"
        }
      ],
      "type": "type/DateTime"
    },
    "19": {
      "name": "Hour of Day",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "hour-of-day"
        }
      ],
      "type": "type/DateTime"
    },
    "20": {
      "name": "Day of Week",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day-of-week"
        }
      ],
      "type": "type/DateTime"
    },
    "21": {
      "name": "Day of Month",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day-of-month"
        }
      ],
      "type": "type/DateTime"
    },
    "22": {
      "name": "Day of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "day-of-year"
        }
      ],
      "type": "type/DateTime"
    },
    "23": {
      "name": "Week of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "week-of-year"
        }
      ],
      "type": "type/DateTime"
    },
    "24": {
      "name": "Month of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "month-of-year"
        }
      ],
      "type": "type/DateTime"
    },
    "25": {
      "name": "Quarter of Year",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "quarter-of-year"
        }
      ],
      "type": "type/DateTime"
    },
    "26": {
      "name": "Minute",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "minute"
        }
      ],
      "type": "type/Time"
    },
    "27": {
      "name": "Hour",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "hour"
        }
      ],
      "type": "type/Time"
    },
    "28": {
      "name": "Minute of Hour",
      "mbql": [
        "field",
        null,
        {
          "temporal-unit": "minute-of-hour"
        }
      ],
      "type": "type/Time"
    },
    "29": {
      "name": "Auto bin",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "default"
          }
        }
      ],
      "type": "type/Number"
    },
    "30": {
      "name": "10 bins",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "num-bins",
            "num-bins": 10
          }
        }
      ],
      "type": "type/Number"
    },
    "31": {
      "name": "50 bins",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "num-bins",
            "num-bins": 50
          }
        }
      ],
      "type": "type/Number"
    },
    "32": {
      "name": "100 bins",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "num-bins",
            "num-bins": 100
          }
        }
      ],
      "type": "type/Number"
    },
    "33": {
      "name": "Don't bin",
      "mbql": null,
      "type": "type/Number"
    },
    "34": {
      "name": "Auto bin",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "default"
          }
        }
      ],
      "type": "type/Coordinate"
    },
    "35": {
      "name": "Bin every 0.1 degrees",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "bin-width",
            "bin-width": 0.1
          }
        }
      ],
      "type": "type/Coordinate"
    },
    "36": {
      "name": "Bin every 1 degree",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "bin-width",
            "bin-width": 1
          }
        }
      ],
      "type": "type/Coordinate"
    },
    "37": {
      "name": "Bin every 10 degrees",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "bin-width",
            "bin-width": 10
          }
        }
      ],
      "type": "type/Coordinate"
    },
    "38": {
      "name": "Bin every 20 degrees",
      "mbql": [
        "field",
        null,
        {
          "binning": {
            "strategy": "bin-width",
            "bin-width": 20
          }
        }
      ],
      "type": "type/Coordinate"
    },
    "39": {
      "name": "Don't bin",
      "mbql": null,
      "type": "type/Coordinate"
    }
  },
  "updated_at": "2023-03-23T21:50:00.139838Z",
  "active": true,
  "id": 9807,
  "db_id": 2177,
  "visibility_type": null,
  "field_order": "database",
  "initial_sync_status": "complete",
  "display_name": "Products",
  "metrics": [],
  "created_at": "2023-03-23T18:04:15.608302Z",
  "points_of_interest": "Is it? You tell us!"
}
```

So it looks like we should include unit options as well -- at least for filter and aggregation

**Comment by camsaul** [comment]
Implemented by #29580
